Conversion between image formats in tomography GUI

FedeMPouzols opened the pull request:
New tab in the tomography GUI. This is a utility to convert between image formats.
Some input test stacks come in different formats, and sometimes users need to convert stacks in FITS or other formats to a different format before loading them into third party software for reconstruction/visualization/segmentation/etc.

This depends on: 
 - [x] #16020, which should be merged first.

**To test:**

- Check the code
- New unit tests should pass
- Check that the documentation reads well.
- For a functional test, grab these files: (ISIS network) `\\olympic\babylon5\Public\fedemp\IMAT\convert_images_dummy_test_example`
- Convert the files to your favorite output format (from the list that the interface supports). For example:
  - Open the custom interface: "Diffraction" => "Tomographic Reconstruction". Tab Convert. Select "Pick Format" as "FITS. Then point the input path to the "dummy_larmor" subdirectory of the test data. Select output "Format" as "png". You can leave the maximum search depth as 3 by default. Enter an output path. See that it generates output images for the input images, reconstructing the same tree of subdirectories in the outputs. The interface will pop up a message when the conversion is finished. You can also use input path "dummy_stfc", input format "png", output format "tiff".
  - Open/close the interface and see that the settings are remembered between sessions.
  - NXTomo format is disabled as input, as we don't have a LoadNXTomo and we may not add it unless there's interest in it.
  - The interface can become unresponsive if the amount of images is large. This is unfortunately as it stands now, but this covers current needs and I wouldn't fix this with a temporary solution but rather add the Load/SaveImage algorithms, see more comments below.
- If you're an aggressive tester, you can check that if you create a directory inside the input path and use it as output path, the interface will not fall in the trap and will  not produce outputs for the output path itself.
- Still, if the input and output path are the same, the output files will be generated "in-place" next to the input files.
- You could also check that if you select an input or output path, and remove one or both of them before clicking "convert", the interface will behave well.

Fixes #15942
Also fixes #11817

(The second ticket was a long standing one about documentation, and now that all the tabs of the interface are documented I'd close it). We'll need one to add a worked out example or similar, which cannot be done at the moment.

For normal image formats (PNG, TIFF, etc.) this uses Qt classes (QImageWriter, QImageReader), which has some limitations but is the temporary solution until we have algorithms `Load/SaveImage` (#6843)

Load/Saving of FITS files is well testes in unit and system tests. NXTomo saving is also tested. Saving/loading other formats is unfortunately not well tested, as this is being done via Qt classes at the moment. The next step before we can improve on that is to have Load/SaveImage algorithms.

Release notes: https://github.com/mantidproject/mantid/commit/253929fd99b20ee2475f1f3d004b456cda08f254


---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the coding standards? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [x] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

AntonPiccardoSelg commented on MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Tomography/ImggFormats.h:
```diff
@@ -0,0 +1,54 @@
+Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+namespace ImggFormats {
+enum Format { FITS = 0, TIFF = 1, PNG = 2, JPG = 3, NXTomo = 4 };
```
is an enum class possible or would you have to convert often in the GUI?

FedeMPouzols replied:
I'd tend to prefer an enum class, but I was being lazy until we introduce the Load/SaveImage algorithms. When we know what underlying library is used for handling the image formats it will be more clear what would be more convenient. For now this is specific to the `Tomography/` sources.

AntonPiccardoSelg commented on docs/source/interfaces/Tomographic_Reconstruction.rst:
```diff
@@ -385,23 +385,39 @@ algorithm processes the input directories and files.
 -------
 
-TODO: there should be an example using a small data set.
+TODO: there should be a worked out example using a small data set.
```
Is this for another PR?

FedeMPouzols replied:
Yes, definitely. There's several `TODO`s here that need further work as soon as we get data, and have a bit more experience with some of the tools so we can fill this in.

AntonPiccardoSelg commented on Framework/DataHandling/src/LoadFITS.cpp:
```diff
@@ -32,9 +32,29 @@ template <typename InterpretType> double toDouble(uint8_t *src) {
+  std::string filePath;
+  bool isFloat;
+};
+
```
good move into the `.cpp`

AntonPiccardoSelg commented on Framework/DataHandling/src/LoadFITS.cpp:
```diff
@@ -150,170 +170,175 @@ void LoadFITS::exec() {
+  try {
+    std::string tmpBitPix = header.headerKeys[m_headerBitDepthKey];
+    if (boost::contains(tmpBitPix, "-")) {
+      boost::erase_all(tmpBitPix, "-");
```
could use `std::find` and `std::replace` but this is actually easier to read :+1:

AntonPiccardoSelg commented on Framework/DataHandling/src/LoadFITS.cpp:
```diff
@@ -150,170 +170,175 @@ void LoadFITS::exec() {
+    // Various extensions to the FITS format are used elsewhere, and
+    // must be parsed differently if used. This loader Loader
+    // doesn't support this.
+    header.extension = header.headerKeys["XTENSION"];
```
should `XTENSION` be defined somewhere as a `static const`?

AntonPiccardoSelg commented on Framework/DataHandling/src/LoadFITS.cpp:
```diff
@@ -150,170 +170,175 @@ void LoadFITS::exec() {
       try {
-        headers[i].scale = boost::lexical_cast<double>(
-            headers[i].headerKeys[m_headerScaleKey]);
+        double doff =
```
could use `auto`

AntonPiccardoSelg commented on Framework/DataHandling/src/SaveFITS.cpp:
```diff
@@ -0,0 +1,252 @@
+                                    std::ofstream &file) {
+  static const std::vector<char> blanks(g_maxLenHdr, 32);
+
+  size_t count = hdr.size();
```
`auto`

AntonPiccardoSelg commented on MantidQt/CustomInterfaces/src/Tomography/ImggFormatsConvertPresenter.cpp:
```diff
@@ -0,0 +1,315 @@
+      boost::dynamic_pointer_cast<Mantid::API::MatrixWorkspace>(
+          grp->getItem(grp->size() - 1));
+
+  return imgWorkspace;
```
Why do you go through the ADS here? Is the workspace used for something else?

FedeMPouzols replied:
Now that you mention this I'm going to double check. I think here it should be possible to do it the neat way with `getProperty()`.

AntonPiccardoSelg commented on MantidQt/CustomInterfaces/src/Tomography/ImggFormatsConvertViewQtWidget.cpp:
```diff
@@ -0,0 +1,376 @@
+    const std::string &outFormat) const {
+  if (!inWks)
+    return;
+  size_t width = inWks->getNumberHistograms();
```
`auto`, also below

AntonPiccardoSelg commented:
Code review:
 * Looks good. Comments are minor and don't warrant a change now.
 * Good unit tests are in place
 * Very extensive documentation is provided, making it clear how the scientists can use the converter.

AntonPiccardoSelg commented:
Funcitonal testing:
 * No pop up appears when the conversion has completed.
 * Converted from FITS to TIFF without any issues
 * Converted from FITS to JPEG --> Mantid terminates (Unexpected exception: map::at)
 * Converted from FITS to PNG without any issues
 * Converted from FITS to NXTOMO --> Mantid terminates (Unexpected exception: All arguments to init must be positive and non-zero)
   
The output files end with `.fits.tiff` and `fits.png`. Is this on purpose?

FedeMPouzols commented:
@AntonPiccardoSelg :  great that you caught these issues in time.
- The pop-up should show up now, I must have downgraded to a log message without noticing. This is now logging the message and also popping up.
- Saving to JPEG and NXTomo should be working well now. The JPEG issue was a silly mistake after an update in the name of the format. The NXTomo issue uncovered a bug in SaveNXTomo algorithm for which I've added a regression test.
- I've added better `try/catch` and more specific error messages for the scenarios where the load/save could fail. That should prevent those panic messages.
- And yes, the original extension is kept as it is useful information in practice. For example the images that were .fits original are most likely coming from the instrument cameras. As not all the image formats support headers this is useful in practice so it is better not to remove the old extension.

AntonPiccardoSelg commented:
Functional Testing:
* Converted from FITS to TIFF --> works well ( I assume that is only for compression)
* Converted from FITS to TIFF --> works well
* Converted from FITS to JPEG --> works well
* Converted from FITS to PNG --> works well
* Converted from FITS to NXTOMO --> works well

* Converted from PNG to PNG --> works well
* Converted from PNG to TIFF --> works well
* Converted from PNG to JPEG --> works well

* Converted from TIFF to PNG --> works well
* Converted from TIFF to TIFF --> works well
* Converted from TIFF to JPEG --> works well

* Converted from JPEG to PNG --> works well
* Converted from JPEG to TIFF --> works well
* Converted from JPEG to JPEG --> works well

* Converting from TIFF/PNG/JPEG to NXTOMO/FITS produces an error pop up (since this conversion is not well defined). It might make sense to disable it in the future.

Happy to :shipit:
